modules/axmm/src/aspace.rs

```diff
@@ -0,0 +1,139 @@
+use core::fmt;
+
+use axerrno::{ax_err, AxError, AxResult};
+use axhal::paging::{MappingFlags, PageTable};
+use memory_addr::{is_aligned_4k, PhysAddr, VirtAddr, VirtAddrRange};
+
+use crate::paging_err_to_ax_err;
+
+/// The virtual memory address space.
+pub struct AddrSpace {
+    va_range: VirtAddrRange,
+    pt: PageTable,
+}
+
+impl AddrSpace {
+    /// Returns the address space base.
+    pub const fn base(&self) -> VirtAddr {
+        self.va_range.start
+    }
+
+    /// Returns the address space end.
+    pub const fn end(&self) -> VirtAddr {
+        self.va_range.end
+    }
+
+    /// Returns the address space size.
+    pub const fn size(&self) -> usize {
+        self.va_range.size()
+    }
+
+    /// Returns the reference to the inner page table.
+    pub const fn page_table(&self) -> &PageTable {
+        &self.pt
+    }
+
+    /// Returns the root physical address of the inner page table.
+    pub const fn page_table_root(&self) -> PhysAddr {
+        self.pt.root_paddr()
+    }
+
+    /// Checks if the address space contains the given address range.
+    pub const fn contains_range(&self, start: VirtAddr, size: usize) -> bool {
+        self.va_range
+            .contains_range(VirtAddrRange::from_start_size(start, size))
+    }
+
+    /// Creates a new empty address space.
+    pub(crate) fn new_empty(base: VirtAddr, size: usize) -> AxResult<Self> {
+        Ok(Self {
+            va_range: VirtAddrRange::from_start_size(base, size),
+            pt: PageTable::try_new().map_err(|_| AxError::NoMemory)?,
+        })
+    }
+
+    /// Add a new linear mapping.
+    ///
+    /// The mapping is linear, i.e., `start_vaddr` is mapped to `start_paddr`,
+    /// and `start_vaddr + size` is mapped to `start_paddr + size`.
+    ///
+    /// The `flags` parameter indicates the mapping permissions and attributes.
+    ///
+    /// Returns an error if the address range is out of the address space or not
```

Comment:
delete "or not"

Reply:
Returns an error if the address range is out of the address space or with bad alignment.